packages/components/src/text-control/types.ts

```diff
@@ -3,7 +3,22 @@
  */
 import type { BaseControlProps } from '../base-control/types';
 
-export type TextControlProps = Pick<
+export type TextControlPropsType =
+	| 'date'
+	| 'datetime-local'
+	| 'email'
+	| 'number'
+	| 'password'
+	| 'tel'
+	| 'text'
+	| 'time'
+	| 'search'
+	| 'url';
+
+export type TextControlProps<
+	Type extends TextControlPropsType = 'text',
+	Value extends string | number = Type extends 'number' ? number : string,
```

Comment:
It might be better to make this a generic type (e.g. `ValueFromType<T>` and then use it in place of the `NoInfer<Value>`'s. The value type is not really a type argument, but a computed type from "Type", which is the actual type argument.

This would also allow passing a Type explicitly without the need to pass a Value (there's no partial inference). Why you'd want to do that? I have no idea. But it's possible, so...

Plus this allows an invalid combination, e.g. type text and value number. It could be fixed by repeating the condition in the extends clause, but I think it's cleaner to just compute it in each prop with a utility generic type as I suggested.

Reply:
updated